### ISSUE TITLE
How to use cy.ntlm when the host that has authentication has a path?

### ISSUE THREAD

**Issue by SheKnowsItsAMultipass** [bjowes/cypress-ntlm-auth#22]
This plugin has been a life saver (Thanks!) but we had this running a certain way but after an update it seems its not possible anymore.  The app that has authentication is on https://someIntranetServer/TheAppWithAuthentication but now the command give an error that I can now see in the readme. "must NOT include the rest of the url (path and query) - only host level authentication is supported"

   Unless I am mistaken a recent change made this impossible.   (I could be wrong about this since we had boat load infrastructure integration and dependency issues for over a week that hid this problem but tests were working fine a couple of weeks ago.)  

If its a new change is it something that could be overridden?  Is there a technical reason? 

I'll admit that my knowledge in all of this is rather limited but bare with me...

I think we might get away with activating authentication on https://someIntranetServer but I can see situations where this may not be possible. If I am testing https://someIntranetServer/TheAppWithAuthentication  and that I can only guaranty the state of this app then authentication shouldn't the authentication be done there?

Hope you can help with some insight.  Thanks!

**Comment by bjowes** [comment]
Hi @SheKnowsItsAMultipass, (lovely alias by the way!)

Thanks for the praise! 

The validation of the url in cy.ntlm was introduced in 0.7.0 as a mitigation to #12. Under the surface, the ntlm-proxy will match all incoming requests to registered NTLM hosts. If there is a match, NTLM authentication is applied, otherwise the request just passes through. 

For pure http, incoming requests were always matched to registered NTLM hosts using only the host part of the url. But for https, it was not (before 0.7.0). I decided it was best to make the behaviour identical for both http and https.
The issue with matching on url with path and query is that the match becomes very specific. Generally, the first request fetches a html page, which triggers a bunch of additional resources (images, scripts) to be fetched with slightly different paths (other filenames). What happened was that fetching those resources failed since the url with path didn't match exactly to the registered NTLM host.

Now, for your described use case, you are fine with registering https://someIntranetServer as the NTLM host, and then accessing only https://someIntranetServer/TheAppWithAuthentication (and deeper paths) with Cypress for your tests. This will work as long as you don't need to access another path on the same server with another domain user (such as https://someIntranetServer/SomeOtherAppWithOtherAuthentication) in the very same test. Note that this includes also requests coming indirectly from the site you are testing (issued from the browser, not directly from Cypress). The reason is that the ntlm-proxy will apply the same authentication to all requests to https://someIntranetServer, regardless of path. 
Also, as long as any other sites (accessed during the test) on https://someIntranetServer don't use NTLM, you will also be fine. You will see some errors if you monitor the logs from ntlm-proxy, but the requests will just pass through when the ntlm-proxy discovers that the server isn't responding to the NTLM handshake.

Let me know if this clarifies your situation.

I do see a potential future feature of this plugin where support for urls with paths is added. It would then assume that all paths that starts with the registered path belongs to the same NTLM host. If you think you will require this, I will look into it.

**Comment by SheKnowsItsAMultipass** [comment]
I dont know if I'm doing somethign wrong.  I still have a problem on my build server that I cant seem to reproduce locally.  And I think my initial assumption was off the mark.  I'm still trying all sorts of things, locally it works as expected but the same config on the dev server gives a 400 bad request error for the ntlm command.

Could it be cause by last weeks cypress update? But why does it seem to work locally when doing my config in the same way as the build server.  I'm still investigating but here is the error  I get.  

  > 400: Bad Request
This was considered a failure because the status code was not '2xx' or '3xx'.
If you do not want status codes to cause failures pass the option: 'failOnStatusCode: false'

-----------------------------------------------------------

The request we sent was:
Method: POST

URL: http://127.0.0.1:21422/ntlm-config

Headers: {

  "user-agent": "Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Cypress/3.1.5 Chrome/59.0.3071.115 Electron/1.8.2 Safari/537.36",

  "accept": "*/*",

  "accept-encoding": "gzip, deflate",

  "content-type": "application/json",

  "content-length": 101

}

Body: {"ntlmHost":"http://Some-server-name","username":"SomeTestUser-001","password":"SomePassword","domain":"MyDomain"

when I paste the url from the body in my browser it loads up fine.

**Comment by bjowes** [comment]
That is odd @SheKnowsItsAMultipass . The current version of the plugin uses the same validation of the ntlm-config on the client and server side - meaning that if the config would be considered invalid by the ntlm-proxy, it would fail before being sent by the cy.ntlm command.
If there is some glitch in this logic, the ntlm-config should at least respond with the validation error in the body of the 400 response. Can you see the response body?

Either way, since you are experiencing different behavior on the build server vs locally, I would really recommend enabling debug logging. This is documented on the main page, but in short you would set the environment variable DEBUG to * before starting ntlm-proxy. Setting it to * enables debug logging for all modules used by the ntlm-proxy, which could prove useful if the ntlm-config request gets stuck on the way and never reaches the validation code.

**Comment by SheKnowsItsAMultipass** [comment]
Well after messing around quite a bit I went on the build server with my account, cleaned out all the build folders, rebooted everything but the build tests would still fail.  Finally still with my account I ran exactly the same commands as the build but had to install cypress manually for my profile and then the cypress tests passed.  And it seems to have fixed the build too =/.  I added a cypress install command to my build script just in case.

Thx for the help, This issue can die.

**Comment by bjowes** [comment]
Glad to hear you solved it @SheKnowsItsAMultipass .
Just for reference, which version of cypress did you have installed on the build server that was consistently failing with my plugin? I could add some minimum version in peerDependencies.

**Comment by SheKnowsItsAMultipass** [comment]
I'm really confused at this point and my bad memory and lack of sleep dont help but I dont feel like ntlm-auth was at fault at this point.

We had cypress 3.1 locally and on the build server.  Installed it like this locally npm install cypress --save-dev and not sure how it was installed on the build server (npm install of our package.json in the jenkins pipeline or manual install previously)  and everythign was working.

Then some day a couple of weeks ago, I think after I did a npm install on our project.  I couldnt get the cypress tests to run on my machine anymore, so I reinstalled cypress and that was version 3.1.5.

I feel this is when the tests stopped working on the build server but that was hidden because they were already failing because the backend server we were hitting was down for a few days. 

Hope this helps paint a better picture.
